akka-http/src/main/scala/akka/http/javadsl/server/directives/PathDirectives.scala

```diff
@@ -1,187 +1,59 @@
-   *     pathPrefix("users") {
-   *       pathEnd {
-   *         // user list ...
-   *       } ~
-   *       path(UUID) { userId =>
-   *         // user profile ...
-   *       }
-   *     }
-   *   }
-   * }
-   * }}}
-   *
-   * For further information, refer to:
-   * [[http://googlewebmastercentral.blogspot.de/2010/04/to-slash-or-not-to-slash.html]]
-   */
-  def pathEndOrSingleSlash: Directive = RawPathPrefixForMatchers(List(PathMatchers.SLASH.optional, PathMatchers.END))
-
-  /**
-   * Only passes on the request to its inner route if the request path
-   * consists of exactly one remaining slash.
-   */
-  def pathSingleSlash: Directive = RawPathPrefixForMatchers(List(PathMatchers.SLASH, PathMatchers.END))
-
-  /**
-   * If the request path doesn't end with a slash, redirect to the same uri with trailing slash in the path.
-   *
-   * '''Caveat''': [[path]] without trailing slash and [[pathEnd]] directives will not match inside of this directive.
-   */
-  @varargs
-  def redirectToTrailingSlashIfMissing(redirectionStatusCode: StatusCode, innerRoute: Route, moreInnerRoutes: Route*): Route =
-    RedirectToTrailingSlashIfMissing(redirectionStatusCode)(innerRoute, moreInnerRoutes.toList)
-
-  /**
-   * If the request path ends with a slash, redirect to the same uri without trailing slash in the path.
-   *
-   * '''Caveat''': [[pathSingleSlash]] directive will not match inside of this directive.
-   */
-  @varargs
-  def redirectToNoTrailingSlashIfPresent(redirectionStatusCode: StatusCode, innerRoute: Route, moreInnerRoutes: Route*): Route =
-    RedirectToNoTrailingSlashIfPresent(redirectionStatusCode)(innerRoute, moreInnerRoutes.toList)
-
-  private def RawPathPrefixForMatchers(matchers: immutable.Seq[PathMatcher[_]]): Directive =
-    Directives.custom(RouteStructure.RawPathPrefix(matchers))
-
-  private def RawPathPrefixTestForMatchers(matchers: immutable.Seq[PathMatcher[_]]): Directive =
-    Directives.custom(RouteStructure.RawPathPrefixTest(matchers))
-
-  private def joinWithSlash(matchers: immutable.Seq[PathMatcher[_]]): immutable.Seq[PathMatcher[_]] = {
-    def join(result: immutable.Seq[PathMatcher[_]], next: PathMatcher[_]): immutable.Seq[PathMatcher[_]] =
-      result :+ PathMatchers.SLASH :+ next
-
-    matchers.foldLeft(immutable.Seq.empty[PathMatcher[_]])(join)
+package akka.http.javadsl.server.directives
+
+import akka.http.scaladsl.server.PathMatchers
+import java.util.function.Supplier
+import akka.http.javadsl.server.Route
+import akka.http.scaladsl.server.PathMatcher.segmentStringToPathMatcher
+import akka.http.scaladsl.server.{ Directives ⇒ ScalaDirectives }
+import akka.http.javadsl.server.Unmarshaller
+import scala.util.Success
+import scala.util.Failure
+import akka.http.scaladsl.server.ValidationRejection
+
+/**
+ * Only path prefixes are matched by these methods, since any kind of chaining path extractions
+ * in Java would just look cumbersome without operator overloads; hence, no PathMatcher for Java.
```

Comment:
I think this is a viable way around the overloading boilerplate madness that would otherwise ensue. It makes Java routes a bit more verbose than their Scala equivalent, but that cost is due to the host language.

Users should be able to compose and reuse these, right?